File: frontend/src/services/businessService.js

import axios from 'axios';

const API_BASE = '/api/business';

/**
 * Fetches all businesses.
 * @returns {Promise<Array>} List of Business objects.
 */
export const getAllBusinesses = async () => {
    const res = await axios.get(`${API_BASE}`);
    return res.data.data; // assuming GenericResponse<T>.data holds payload
};

/**
 * Fetches a single business by ID.
 * @param {string} id UUID of the business.
 * @returns {Promise<Object>} Business object.
 */
export const getBusinessById = async (id) => {
    const res = await axios.get(`${API_BASE}/${id}`);
    return res.data.data;
};

/**
 * Searches businesses by name (case-insensitive).
 * @param {string} name Search term.
 * @returns {Promise<Array>} Matched businesses.
 */
export const searchBusinesses = async (name) => {
    const res = await axios.get(`${API_BASE}/search`, { params: { name } });
    return res.data.data;
};

/**
 * Retrieves businesses owned by a specific owner.
 * @param {number} ownerId Owner's numeric ID.
 * @returns {Promise<Array>} List of businesses.
 */
export const getBusinessesByOwner = async (ownerId) => {
    const res = await axios.get(`${API_BASE}/owner/${ownerId}`);
    return res.data.data;
};

/**
 * Retrieves top-rated businesses with a limit.
 * @param {number} limit Number of top items.
 * @returns {Promise<Array>} List of top-rated businesses.
 */
export const getTopRated = async (limit = 5) => {
    const res = await axios.get(`${API_BASE}/top`, { params: { limit } });
    return res.data.data;
};

/**
 * Fetches businesses tagged with a specific UUID.
 * @param {string} tagId UUID string of tag.
 * @returns {Promise<Array>} List of businesses.
 */
export const getByTag = async (tagId) => {
    const res = await axios.get(`${API_BASE}/tag/${tagId}`);
    return res.data.data;
};

export const getBusinessReviews = async (businessId) => {
    const res = await axios.get(`${API_BASE}/reviews/${businessId}`);
    return res.data.data;
};

export const getBusinessPromotions = async (businessId) => {
    const res = await axios.get(`${API_BASE}/promotions/${businessId}`);
    return res.data.data;
}

export const newPromotion = async (businessId, newPromotionData) => {
    const promotion = {
        title: newPromotionData.title,
        description: newPromotionData.description,
        startDate: newPromotionData.startDate,
        endDate: newPromotionData.endDate,
        amount: newPromotionData.amount,
        isActive: newPromotionData.isActive
    };

    const res = await axios.post(`${API_BASE}/promotions/${businessId}`, promotion);
    return res.data.data;
}
export const updatePromotion = async (businessId, promotionId, newPromotionData) => {
    const promotion = {
        title: newPromotionData.title,
        description: newPromotionData.description,
        startDate: newPromotionData.startDate,
        endDate: newPromotionData.endDate,
        amount: newPromotionData.amount,
        isActive: newPromotionData.isActive
    };

    const res = await axios.patch(`${API_BASE}/promotions/${businessId}/${promotionId}`, promotion);
    return res.data.data;
}

export const deletePromotion = async (businessId, promotionId) => {
    const res = await axios.delete(`${API_BASE}/promotions/${businessId}/${promotionId}`);
    return res.data.data;
}

export const setViewed = async (reviewId) => {
    const res = await axios.patch(`${API_BASE}/reviews/${reviewId}`);
    return res.data.data;
}

File: IOS/Services/BusinessService.swift
import Foundation

/// Service responsible for fetching business related data from backend via `APIClient`.
final class BusinessService: @unchecked Sendable {
    private let api: APIClientProtocol
    private let base = "api/business"

    init(api: APIClientProtocol = APIClient.shared) {
        self.api = api
    }

    func getAllBusinesses() async throws -> [Restaurant] {
        let dtos: [BusinessDTO] = try await api.request(base)
        return dtos.map(BusinessMapper.map)
    }

    func getBusinessById(_ id: String) async throws -> Restaurant {
        let dto: BusinessDTO = try await api.request("\(base)/\(id)")
        return BusinessMapper.map(dto)
    }

    func searchBusinesses(_ name: String) async throws -> [Restaurant] {
        let encoded = name.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? name
        let dtos: [BusinessDTO] = try await api.request("\(base)/search?name=\(encoded)")
        return dtos.map(BusinessMapper.map)
    }

    func getBusinessesByOwner(_ ownerId: String) async throws -> [Restaurant] {
        let dtos: [BusinessDTO] = try await api.request("\(base)/owner/\(ownerId)")
        return dtos.map(BusinessMapper.map)
    }

    func getTopRated(limit: Int = 5) async throws -> [Restaurant] {
        // Backend GenericResponse wrapper kullanıyor, APIClient otomatik unwrap yapacak
        let dtos: [BusinessDTO] = try await api.request("\(base)/top?limit=\(limit)")
        return dtos.map(BusinessMapper.map)
    }

    func getByTag(_ tagId: String) async throws -> [Restaurant] {
        let dtos: [BusinessDTO] = try await api.request("\(base)/tag/\(tagId)")
        return dtos.map(BusinessMapper.map)
    }

    func getBusinessReviews(_ businessId: String) async throws -> [Review] {
        let dtos: [ReviewInfoForBusinessDTO] = try await api.request("\(base)/reviews/\(businessId)")
        return dtos.map(ReviewInfoMapper.map)
    }

    func getBusinessPromotions(_ businessId: String) async throws -> [Promotion] {
        let dtos: [PromotionDTO] = try await api.request("\(base)/promotions/\(businessId)")
        return dtos.map(PromotionMapper.map)
    }

    func getNearby(latitude: Double, longitude: Double) async throws -> [Restaurant] {
        let dtos: [BusinessDTO] = try await api.request("\(base)/nearby?lat=\(latitude)&lng=\(longitude)")
        return dtos.map(BusinessMapper.map)
    }

    func newPromotion(_ businessId: String, promotion: PromotionRequest) async throws -> Promotion {
        let body = try encodePromotionRequest(promotion)
        let dto: PromotionDTO = try await api.request("\(base)/promotions/\(businessId)", method: "POST", body: body)
        return PromotionMapper.map(dto)
    }

    func updatePromotion(_ businessId: String, promotionId: String, promotion: PromotionRequest) async throws {
        let body = try encodePromotionRequest(promotion)
        try await api.request(
            "\(base)/promotions/\(businessId)/\(promotionId)",
            method: "PATCH",
            body: body
        ) as EmptyResponse
        if let message = api.message {
            print(message)
        }
    }

    func deletePromotion(_ businessId: String, promotionId: String) async throws {
        try await api.request(
            "\(base)/promotions/\(businessId)/\(promotionId)",
            method: "DELETE",
            body: nil
        ) as EmptyResponse
        if let message = api.message {
            print(message)
        }
    }

    func setViewed(_ reviewId: String) async throws -> EmptyResponse {
        return try await api.request("\(base)/reviews/\(reviewId)", method: "PATCH", body: nil)
    }
}

struct PromotionRequest: Encodable {
    let title: String
    let description: String?
    let startDate: String?
    let endDate: String?
    /// Amount is now an optional integer. Encode as `Int` to avoid floating point
    /// representations.
    let amount: Int?
    let isActive: Bool

    private enum CodingKeys: String, CodingKey {
        case title, description, startDate, endDate, amount, isActive
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(title, forKey: .title)
        try container.encodeIfPresent(description, forKey: .description)
        try container.encodeIfPresent(startDate, forKey: .startDate)
        try container.encodeIfPresent(endDate, forKey: .endDate)
        if let amount = amount {
            try container.encode(Int(amount), forKey: .amount)
        }
        try container.encode(isActive, forKey: .isActive)
    }
}

private func encodePromotionRequest(_ promotion: PromotionRequest) throws -> Data {
    let request = PromotionRequest(
        title: promotion.title,
        description: promotion.description,
        startDate: promotion.startDate,
        endDate: promotion.endDate,
        amount: promotion.amount,
        isActive: promotion.isActive
    )
    return try JSONEncoder().encode(request)
}

